---
title: "Identifying GOI Potato"
author: "Guillian Hern√°ndez"
date: "2024-06-05"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

## Objective

The objective of this Readme is to show how GOI for potato, as well as
TF were determined for the S. Tuberosum Atlantic cv.

## Identifying GOI

Based off the literature, I created a file:
[GHC_Potato_GOI_DM_annotation.csv]{style="color:red"} that has the GOI,
reference paper, and how I was able to find the locus id for DM.\
I now need to make:\
1) File that contains the CDS sequence from the GOI in DM genome, which
I will be searching for in the ATL genome:
[GHC_GOI_DM_cds_sequences.fa]{style="color:red"}\
2) File that contains one column the GOI_ID and in the other, a common
name to use for that gene id : [GHC_GOI_DM_Loci.csv]{style="color:red"}

## Identifying TF

For the TF, I used the Plant Transcription Factor
Database(<https://planttfdb.gao-lab.org/download.php>) to download the\
CDS:[GHC_PSGC_TF_cds.fas]{style="color:red"}\
TF list:[GHC_PSGC_TF_list.txt]{style="color:red"}\
The original file has three columns: TF ID, Gene_ID, Family. Im going to
use the TF_ID and the Family as the annotation for the TF. These are
both for PSGC genome hence, I need to run my pipeline to find the
corresponding genes in Atlantic.

## Peptide Files for Genomes

Now that I have the TF and the GOI I'm intersted in finding, I will run
my pipeline to find orhtologs from the PSGC genome (TF) and the DM
genome(GOI) in Atlantic genome and properly annotate them based on the
identifiers from the orthologus gene in either PSGC or DM genome.
Peptide files were all the representative proteome and were downloaded
from SpuddDB.\

Atlantic:[ATL_v3.working_models.repr.pep.fa]{style="color:red"}\
DM:[DM_1-3_516_R44_potato.v6.1.repr_hc_gene_models.pep.fa]{style="color:red"}\
PSGC: [PGSC_DM_v3.4_pep_representative.fasta]{style="color:red"}

## CDS Atlantic

We need the Coding sequence for Atlantic so that when Blastn is ran, it
can search for the quuery sequences from GOI and TF in the Atlantic
genome cds. Again, I'm using the representative cds file.

Atlantic CDS:[ATL_v3.working_models.repr.cds.fa]

## Setting Directories

Since TF and GOI are using two different reference Genomes, I will make
two directories, one for identifying TF and the other for GOI.

For TF:\
**Home**\
- Genes_of_interest.fa = GHC_PSGC_TF_cds.fas\
- CDS_genome_b.fa = ATL_v3.working_models.repr.cds.fa\
- Gene_annotations.txt = GHC_PSGC_TF_list.txt\
**Home/ortho-genomes**\
- Peptide_genome_a.fa =PGSC_DM_v3.4_pep_representative.fasta\
- Peptide_genome_b.fa =ATL_v3.working_models.repr.pep.fa\

For GOI:\
**Home**\
- Genes_of_interest.fa = GHC_GOI_DM_cds_sequences.fa\
- CDS_genome_b.fa = ATL_v3.working_models.repr.cds.fa\
- Gene_annotations.txt = GHC_GOI_DM_Loci.txt\
**Home/ortho-genomes**\
-
Peptide_genome_a.fa=DM_1-3_516_R44_potato.v6.1.repr_hc_gene_models.pep.fa\
- Peptide_genome_b.fa =ATL_v3.working_models.repr.pep.fa\

Code used for GOI (Different to that of TF).

```{bash}
#!/bin/bash

#SBATCH -J pipeline_GOI # change this with the name of your job
#SBATCH --nodes=1
#SBATCH --mem=64G

# Load OrthoFinder module
module load OrthoFinder/2.5.4-foss-2020b

# Run OrthoFinder
orthofinder -f ./ortho-genomes/

# Load BLAST+ module
odule load BLAST+/2.13.0-gompi-2022a

# Create BLAST database
makeblastdb -in CDS_genome_b_[ATL].fa -dbtype nucl -out CDS_genome_b[ATL].fa_db


# Run BLASTN
blastn -query Genes_of_interest.fa -db CDS_genome_b[ATL].fa_db -perc_identity 95  -evalue 1e-5 -out Ortho+BLASTn_results.txt -outfmt "6 sacc qsedid qacc evalue sallseqid pident ppos"


cp ./ortho-genomes/OrthoFinder/*/Orthogroups/Orthogroups.tsv . 


# Run the Python script for further analysis
python3 <<EOF
import pandas as pd

# Paths to the input files
orthogroups_path = "./Orthogroups.tsv"
GOI_path = "./Ortho+BLASTn_results.txt"
GOI_annotations_path = "./Gene_annotations.txt"

# Read the input files
orthogroups_df = pd.read_csv(orthogroups_path, sep='\t')
GOI_df = pd.read_csv(GOI_path, sep='\t', header=None)
GOI_annotations_df = pd.read_csv(GOI_annotations_path, sep='\t')

# Print first few rows to ensure data is loaded correctly
print("Orthogroups DataFrame:")
print(orthogroups_df.head())

print("\nGOI DataFrame:")
print(GOI_df.head())

print("\nGOI Annotations DataFrame:")
print(GOI_annotations_df.head())

# Rename columns in GOI DataFrame for merging
GOI_df.columns = ['Subject_accession', 'Query_sequence', 'E_value', 'Subject_all_sequence_identifier', 'Percentage_identity', 'Percentage_positives']

# Print DataFrame to check renaming
print("\nRenamed GOI DataFrame:")
print(GOI_df.head())

# Function to split rows in orthogroups DataFrame
def split_orthogroups(df, col_name, new_col_name):
    split_df = df[col_name].str.split(', ', expand=True).stack().reset_index(level=1, drop=True)
    split_df.name = new_col_name
    return df.drop(col_name, axis=1).join(split_df)

# Split the orthogroups DataFrame
orthogroups_df_a = split_orthogroups(orthogroups_df, 'Peptide_genome_a_[DM]', 'Peptide_genome_a')
orthogroups_df_b = split_orthogroups(orthogroups_df, 'Peptide_genome_b_[ATL]', 'Peptide_genome_b')

# Merge the split DataFrames
orthogroups_split_df = pd.merge(orthogroups_df_a, orthogroups_df_b, on='Orthogroup')

print("\nSplit Orthogroups DataFrame:")
print(orthogroups_split_df.head())

# First merge operation
merged_df = pd.merge(GOI_df, orthogroups_split_df, how='left', left_on=['Subject_accession', 'Query_sequence'], right_on=['Peptide_genome_b', 'Peptide_genome_a'])
print("\nMerged DataFrame (left join):")
print(merged_df.head())

# Second merge operation for exact matches
exact_match_df = pd.merge(GOI_df, orthogroups_split_df, how='inner', left_on=['Subject_accession', 'Query_sequence'], right_on=['Peptide_genome_b', 'Peptide_genome_a'])
print("\nExact Match DataFrame (inner join):")
print(exact_match_df.head())

# Save the exact matches to a CSV file
exact_match_df.to_csv('Genome_A_GOIs_in_Genome_B_with_OGs.csv', index=False)

# Merge with GOI annotations
final_GOI_df = pd.merge(exact_match_df, GOI_annotations_df, how='inner', left_on='Query_sequence', right_on='GOI_ID')
print("\nFinal GOI DataFrame after merging with annotations:")
print(final_GOI_df.head())

# Save the final transcription factor results to a CSV file
final_GOI_df.to_csv('GOI_in_genome_b_results.csv', index=False)

# Extract and save the Transcription Factor loci data
final_GOI_loci = final_GOI_df[['Gene', 'Subject_accession']]
final_GOI_loci.to_csv('Final_GOI_in_Genome_b.csv', index=False)
EOF


```

Code for TF:

```{bash}
#!/bin/bash

#SBATCH -J pipeline_TF # change this with the name of your job
#SBATCH --nodes=1
#SBATCH --mem=64G
#!/bin/bash


# Load OrthoFinder module
module load OrthoFinder/2.5.4-foss-2020b

# Run OrthoFinder
#rthofinder -f ./ortho-genomes/

# Load BLAST+ module
module load BLAST+/2.13.0-gompi-2022a

# Create BLAST database
makeblastdb -in CDS_genome_b_[ATL].fa -dbtype nucl -out CDS_genome_b[ATL].fa_db


# Run BLASTN
blastn -query Genes_of_interest.fa -db CDS_genome_b[ATL].fa_db -perc_identity 95  -evalue 1e-5 -out Ortho+BLASTn_results.txt -outfmt "6 sacc qsedid qacc evalue sallseqid pident ppos"


cp ./ortho-genomes/OrthoFinder/*/Orthogroups/Orthogroups.tsv . 


# Run the Python script for further analysis
python3 <<EOF
import pandas as pd

# Paths to the input files
orthogroups_path = "./Orthogroups.tsv"
GOI_path = "./Ortho+BLASTn_results.txt"
GOI_annotations_path = "./Gene_annotations.txt"

# Read the input files
orthogroups_df = pd.read_csv(orthogroups_path, sep='\t')
GOI_df = pd.read_csv(GOI_path, sep='\t')
GOI_annotations_df = pd.read_csv(GOI_annotations_path, sep='\t')

# Print first few rows to ensure data is loaded correctly
print("Orthogroups DataFrame:")
print(orthogroups_df.head())

print("\nGOI DataFrame:")
print(GOI_df.head())

print("\nGOI Annotations DataFrame:")
print(GOI_annotations_df.head())

# Rename columns in GOI DataFrame for merging
GOI_df.columns = ['Subject_accession', 'Query_sequence', 'E_value', 'Subject_all_sequence_identifier', 'Percentage_identity', 'Percentage_positives']

# Print DataFrame to check renaming
print("\nRenamed GOI DataFrame:")
print(GOI_df.head())

# Function to split rows in orthogroups DataFrame
def split_orthogroups(df, col_name, new_col_name):
    split_df = df[col_name].str.split(', ', expand=True).stack().reset_index(level=1, drop=True)
    split_df.name = new_col_name
    return df.drop(col_name, axis=1).join(split_df)

# Split the orthogroups DataFrame
orthogroups_df_a = split_orthogroups(orthogroups_df, 'Peptides_genome_a_[PSGC]', 'Peptide_genome_a')
orthogroups_df_b = split_orthogroups(orthogroups_df, 'Peptides_genome_b_[ATL]', 'Peptide_genome_b')

# Merge the split DataFrames
orthogroups_split_df = pd.merge(orthogroups_df_a, orthogroups_df_b, on='Orthogroup')

print("\nSplit Orthogroups DataFrame:")
print(orthogroups_split_df.head())

# First merge operation
merged_df = pd.merge(GOI_df, orthogroups_split_df, how='left', left_on=['Subject_accession', 'Query_sequence'], right_on=['Peptide_genome_b', 'Peptide_genome_a'])
print("\nMerged DataFrame (left join):")
print(merged_df.head())

# Second merge operation for exact matches
exact_match_df = pd.merge(GOI_df, orthogroups_split_df, how='inner', left_on=['Subject_accession', 'Query_sequence'], right_on=['Peptide_genome_b', 'Peptide_genome_a'])
print("\nExact Match DataFrame (inner join):")
print(exact_match_df.head())

# Save the exact matches to a CSV file
exact_match_df.to_csv('Genome_A_GOIs_in_Genome_B_with_OGs.csv', index=False)

# Merge with GOI annotations
final_GOI_df = pd.merge(exact_match_df, GOI_annotations_df, how='inner', left_on='Query_sequence', right_on='TF_ID')
print("\nFinal GOI DataFrame after merging with annotations:")
print(final_GOI_df.head())

# Save the final transcription factor results to a CSV file
final_GOI_df.to_csv('GOI_in_genome_b_results.csv', index=False)

# Extract and save the Transcription Factor loci data
final_GOI_loci = final_GOI_df[['Family', 'Subject_accession']]
final_GOI_loci.to_csv('Final_GOI_in_Genome_b.csv', index=False)
EOF

```

## Pipeline Output Files

This pipeline outputs two important files:\

1)  Final_GOI_in_Genome_b.csv - Table with loci in genome b and its
    annotation\

2)  GOI_in_genome_b_results.csv - Table with GOIs in genome A, GOIs in
    genome B, and Blastn results Orthogroups.tsv\

3)  Orthogroups.tsv - Table with Orthogroups between Genome A and B\

Files saved as:

GOI:

[GHC_Tuber_GOI_ATL_annotation.csv]{style="color:red"} =
GOI_in_genome_b_results.csv\
[GHC_Tuber_GOI_ATL.csv]{style="color:red"} = Final_GOI_in_Genome_b.csv\
[Orthogroups_ATL-DM.tsv]{style="color:red"} = Orthogroups.tsv

TF: [GHC_STuberosum_TF_annotation.csv]{style="color:red"}=
GOI_in_genome_b_results.csv\
[GHC_STuberosum_TF.csv]{style="color:red"}= Final_GOI_in_Genome_b.csv\
[Orthogroups_ATL-PSGC.tsv]{style="color:red"}= Orthogroups.tsv

#Tuber GOI File

Since I've already described some genes in past manuscripts, I will
merge these two to know which gene is which for personal use in the
future

```{r}
# Set the working directory to the location of the data files
setwd("/Users/gih0004/Desktop/Research/Atl_Genome_Resources_2024/ATLv3_genome/GHC_GOI_Pipeline_Results/ATL-DM_GOI")

# Load the file containing past tuber transcripts from PlantPhys
PlantPhys_GOI <- read.csv("Tuber_Transcripts_PlantPhys.csv", header = TRUE)
colnames(PlantPhys_GOI)

# Load the file containing GHC tuber genes of interest
GHC_GOI <- read.csv("GHC_Tuber_GOI_ATL.csv", header = TRUE)
colnames(GHC_GOI)

# Merge the two data frames on the 'Subject_accession' column from GHC_GOI
# and the 'Transcript_id' column from PlantPhys_GOI, keeping all rows from PlantPhys_GOI
Merged_GOI <- merge(GHC_GOI, PlantPhys_GOI, by.x = "Subject_accession", by.y = "Transcript_id", all.y = TRUE)

# Rename columns to the specified names
colnames(Merged_GOI) <- c("Subject_accession", "Pipeline_Gene", "Manuscript_Gene_name")

# Write the merged data frame to a CSV file
write.csv(Merged_GOI, "GHC_GOI-Manuscript_Gene_ID.csv", row.names = FALSE)


```

Output file: [GHC_GOI-Manuscript_Gene_ID.csv]{style="color:red"}

Now, Since some of my GOI are also TF, I want to know which ones are TF from my piepline so that I can properly identify them visualy in the GRN:
```{r}
setwd("/Users/gih0004/Desktop/Research/Atl_Genome_Resources_2024/ATLv3_genome/GHC_GOI_Pipeline_Results/ATL-DM_GOI")
GOI<-read.csv("GHC_GOI-Manuscript_Gene_ID.csv",header=TRUE)


setwd("/Users/gih0004/Desktop/Research/Atl_Genome_Resources_2024/ATLv3_genome/GHC_GOI_Pipeline_Results/ATL-PSGC_TF")
TF<-read.csv("GHC_Stuberosum_TF.csv",header=TRUE)
TF<-unique(TF)
write.csv(TF,"GHC_Stuberosum_TF.csv")

merged_GOI_TF<-merge(GOI,TF,by="Subject_accession",all.x=TRUE)
write.csv(merged_GOI_TF,"GHC_GOI_Which_are_TF.csv")


```


